File: sdk/cli/src/api/catalog.ts
/**
 * What endpoints exist, and what each one expects.
 *
 * Discovery is the reason this tool is worth building for an agent: without it
 * the caller needs a checkout to know the surface, and the whole premise is a
 * caller who has neither a checkout nor a doc site open.
 *
 * The Rust `oxy api` answered this from data baked into the binary at build
 * time — `crates/app/build_route_catalog.rs` walks the axum router source and
 * emits a table. That has one flaw this design fixes: the table describes the
 * routes the BINARY could mount, and several mounts are mode-conditional, so a
 * listed path can still 404 on the deployment in front of you. Asking the
 * deployment is strictly more truthful.
 *
 * It costs an authenticated network call, so the answer is cached per host and
 * a failure degrades to STALE AND SAID SO rather than to nothing — the
 * discipline `customer-tooling`'s `lib/customers.sh` writes down at length,
 * and for the same reason: an empty answer here reads as "this deployment has
 * no such route", which is a lie that sends the caller off to debug the wrong
 * thing.
 */

import { readFileSync, writeFileSync } from "node:fs";
import { join } from "node:path";
import { hostKey } from "../auth/credentials.js";
import * as log from "../ui/log.js";
import { CliError, ExitCode } from "../util/errors.js";
import { ensureDir, oxycCacheDir, slugifyForPath } from "../util/paths.js";
import { parseJson, request } from "./request.js";

/** One endpoint, as `/api/_catalog` reports it. Mirrors Rust `RouteDescription`. */
export interface CatalogRoute {
  method: string;
  path: string;
  surface: string;
  /** The credential this surface expects, spelled out rather than implied. */
  credential: string;
  path_parameters: string[];
  description: string;
  note: string;
  handler: string;
  /** `ide-only`, `worker-only` or `fleet-ok` — the role_manifest's own answer. */
  role: string;
}

export interface Catalog {
  routes: CatalogRoute[];
  surfaces: Array<{ id: string; label: string; credential: string }>;
  /** Which deployment answered, and when — both shown when serving stale. */
  host: string;
  fetchedAt: number;
  /** True when this came off disk after the network failed. */
  stale?: boolean;
}

/** An hour: long enough that a session costs one call, short enough to notice a deploy. */
const CATALOG_TTL_MS = 3_600_000;

function catalogPath(target: string): string {
  // 0700/0600 like the response cache: a full route table for a deployment
  // is reconnaissance, which is why the endpoint serving it is authenticated.
  const dir = ensureDir(join(oxycCacheDir(), "catalog"), 0o700);
  return join(dir, `${slugifyForPath(hostKey(target))}.json`);
}

function readDisk(target: string): Catalog | undefined {
  try {
    const cached = JSON.parse(readFileSync(catalogPath(target), "utf8")) as Catalog;
    // The filename is a sanitised host and can in principle collide; the
    // stored host is the identity. Serving one deployment's route table as
    // another's would be indistinguishable from a routing bug.
    if (cached.host !== hostKey(target)) return undefined;
    return cached;
  } catch {
    return undefined;
  }
}

function writeDisk(target: string, catalog: Catalog): void {
  try {
    writeFileSync(catalogPath(target), JSON.stringify(catalog), { mode: 0o600 });
  } catch {
    // A cache that cannot be written costs a call per command, not an answer.
  }
}

export interface CatalogOptions {
  target: string;
  bearer?: string;
  /** Ignore a fresh cache and ask the deployment again. */
  refresh?: boolean;
}

/**
 * The route table for `target`.
 *
 * Fresh cache → live `/api/_catalog` → stale cache with a warning → an error
 * that says which of the two things went wrong. Never an empty catalog: an
 * empty list would be reported as "no such route" by every caller.
 *
 * ASKS FOR THE WHOLE TABLE even though the endpoint takes a `?filter=`,
 * because the cache is the point: one full fetch per host per hour serves
 * every later `oxyc routes <anything>` offline, where a filtered fetch would
 * be a round trip per query and could not be cached under one key without
 * poisoning it — a cached "routes matching threads" is not a route table, and
 * serving it as one would report every other endpoint as missing. `?filter=`
 * still earns its place on the endpoint, for the cache-less caller: a script,
 * a `curl`, a one-off.
 */
export async function loadCatalog(opts: CatalogOptions): Promise<Catalog> {
  const cached = readDisk(opts.target);
  if (!opts.refresh && cached && Date.now() - cached.fetchedAt < CATALOG_TTL_MS) {
    return cached;
  }

  // NO CREDENTIAL AND NOTHING CACHED: say so before asking, because the answer
  // an unauthenticated request gets back is not diagnostic. `/api/_catalog` is
  // on the protected router, and an anonymous request to it comes back 404 —
  // which this code would otherwise report as "that deployment predates route
  // discovery", sending the reader off to check a deployment version when what
  // they actually need is to log in. With a cache in hand the stale path below
  // is still better than an error, so the check is only for the empty case.
  if (!opts.bearer && !cached) {
    throw new CliError(`not authenticated for ${opts.target}`, {
      code: ExitCode.AUTH,
      hint: `oxyc login --target ${opts.target}   — route discovery needs a token`
    });
  }

  try {
    const response = await request({
      target: opts.target,
      path: "/api/_catalog",
      method: "GET",
      bearer: opts.bearer,
      timeoutMs: 30_000
    });

    if (response.status === 401 || response.status === 403) {
      if (cached) return serveStale(cached, "not authenticated for the route catalog");
      throw new CliError(`the route catalog at ${opts.target} rejected this token`, {
        code: ExitCode.AUTH,
        hint: `oxyc login --target ${opts.target}`
      });
    }
    if (response.status === 404) {
      // An older deployment that predates `/api/_catalog`. Not an error the
      // caller can act on beyond "this one cannot tell you", so say exactly
      // that rather than "not found", which reads as a bad path.
      if (cached) return serveStale(cached, `${opts.target} has no /api/_catalog (older build)`);
      throw new CliError(`${opts.target} does not serve /api/_catalog`, {
        code: ExitCode.NOT_FOUND,
        hint: "that deployment predates route discovery — try `oxyc openapi` for the documented subset"
      });
    }
    if (response.status < 200 || response.status >= 300) {
      if (cached) return serveStale(cached, `catalog request failed (${response.status})`);
      throw new CliError(`could not read the route catalog (${response.status})`, {
        code: ExitCode.UNAVAILABLE
      });
    }

    const payload = parseJson(response.body) as Partial<Catalog> | undefined;
    if (!payload?.routes?.length) {
      // An EMPTY catalog is refused, never served. It is indistinguishable
      // from a deployment with no routes, and every caller downstream would
      // report a real endpoint as nonexistent.
      if (cached) return serveStale(cached, "the deployment returned an empty route catalog");
      throw new CliError("the deployment returned an empty route catalog", {
        code: ExitCode.UNAVAILABLE,
        hint: "this is a server bug, not a bad path — report it rather than working around it"
      });
    }

    const catalog: Catalog = {
      routes: payload.routes,
      surfaces: payload.surfaces ?? [],
      host: hostKey(opts.target),
      fetchedAt: Date.now()
    };
    writeDisk(opts.target, catalog);
    return catalog;
  } catch (cause) {
    if (cause instanceof CliError && !cached) throw cause;
    if (cached) return serveStale(cached, (cause as Error).message);
    throw cause;
  }
}

/**
 * Serve the cache and say so.
 *
 * The warning is not decoration. A stale route table is usually right and
 * occasionally wrong in the one way that matters — a route added since the
 * cache was written reads as nonexistent — so the reader has to know which
 * kind of answer they are holding.
 */
function serveStale(cached: Catalog, why: string): Catalog {
  const ageMinutes = Math.round((Date.now() - cached.fetchedAt) / 60_000);
  log.warn(`serving a STALE route catalog (${ageMinutes}m old): ${why}`);
  log.hint("oxyc routes --refresh   once the deployment is reachable");
  return { ...cached, stale: true };
}

/** Routes whose method, path or surface contains `needle`, case-insensitively. */
export function searchRoutes(catalog: Catalog, needle: string | undefined): CatalogRoute[] {
  if (!needle) return catalog.routes;
  const n = needle.toLowerCase();
  return catalog.routes.filter(
    (r) =>
      r.path.toLowerCase().includes(n) ||
      r.method.toLowerCase().includes(n) ||
      r.surface.toLowerCase().includes(n) ||
      r.description.toLowerCase().includes(n)
  );
}

/** The OpenAPI document the deployment publishes. */
export async function loadOpenApi(opts: CatalogOptions): Promise<unknown> {
  const response = await request({
    target: opts.target,
    path: "/apidoc/openapi.json",
    method: "GET",
    bearer: opts.bearer,
    timeoutMs: 30_000
  });
  if (response.status < 200 || response.status >= 300) {
    throw new CliError(`could not read the OpenAPI document (${response.status})`, {
      code: ExitCode.UNAVAILABLE,
      hint: `check ${opts.target}/apidoc is served by this deployment`
    });
  }
  const doc = parseJson(response.body);
  if (doc === undefined) {
    throw new CliError("the OpenAPI endpoint did not return JSON", {
      code: ExitCode.UNAVAILABLE
    });
  }
  return doc;
}

File: crates/app/src/server/api/catalog.rs
//! `GET /api/_catalog` — the route table, served by the deployment that owns it.
//!
//! # Why this exists
//!
//! `oxyc` (the TypeScript CLI, `sdk/cli`) is the discovery surface for anyone
//! who has a token and nothing else: no checkout, no doc site, no Swagger UI
//! they can browse. It has to be able to answer "what can I call here?" — and
//! it cannot generate the answer, because the answer comes from a build-time
//! walk of this crate's router source (`build_route_catalog.rs`).
//!
//! The Rust `oxy api --routes` answered it from the table baked into the
//! binary. That has one flaw this endpoint fixes: the baked table describes
//! what the BINARY could mount, and several mounts are mode-conditional
//! (`/setup/*` and the git routes exist only in local mode), so a listed path
//! could still 404 on the deployment in front of you. Asking the deployment
//! is strictly more truthful, and it is the only form of the question that
//! stays correct as a caller moves between local, dev and production.
//!
//! # Why it is authenticated
//!
//! It sits on the protected router. A complete route table for a multi-tenant
//! SaaS — admin surfaces, partner console, billing, every path parameter — is
//! a reconnaissance gift, and there is no caller who needs it before they have
//! a token: the CLI holds one for every other command it runs. The client
//! caches per host, so the cost is one call an hour, not one per command.
//!
//! # Scope
//!
//! Whatever `server::route_catalog` covers: the `/api` and `/external/api`
//! surfaces. See that module for what is deliberately outside those bounds
//! (the custom-app bundle tree, the worker health port, the internal loopback
//! router).

use axum::Json;
use axum::extract::Query;
use serde::{Deserialize, Serialize};

use crate::server::route_catalog::{self, RouteDescription};

/// `?filter=` narrows the table server-side.
#[derive(Deserialize, Default)]
pub struct CatalogQuery {
    /// Substring matched against method, path, surface and description.
    filter: Option<String>,
}

/// The document `oxyc` caches per host.
#[derive(Serialize)]
pub struct CatalogResponse {
    /// Every route on the `/api` and `/external/api` surfaces.
    routes: Vec<RouteDescription>,
    /// Surfaces in display order, each with the credential it expects.
    surfaces: Vec<CatalogSurface>,
    /// The build this table was generated from, so a stale client cache and a
    /// redeployed server can be told apart without guessing from timestamps.
    version: &'static str,
}

#[derive(Serialize)]
pub struct CatalogSurface {
    id: &'static str,
    label: &'static str,
    credential: &'static str,
}

/// The route table for this deployment, with what each endpoint does and which
/// credential its surface expects.
///
/// Consumed by `oxyc routes` / `oxyc schema`; the OpenAPI document at
/// `/apidoc/openapi.json` carries the request/response schemas for the
/// curated subset that has them.
///
/// `?filter=<substring>` narrows it here rather than in the client. The whole
/// table is ~670 routes and a few hundred KB of prose; a caller looking for
/// three of them should not have to download all of it, and a filtered request
/// is also the shape a cache-less caller wants.
pub async fn get_catalog(Query(query): Query<CatalogQuery>) -> Json<CatalogResponse> {
    let needle = query
        .filter
        .as_deref()
        .map(str::trim)
        .filter(|f| !f.is_empty());
    Json(CatalogResponse {
        routes: route_catalog::search(needle)
            .into_iter()
            .map(route_catalog::describe)
            .collect(),
        surfaces: route_catalog::surfaces()
            .iter()
            .map(|(id, label, credential)| CatalogSurface {
                id,
                label,
                credential,
            })
            .collect(),
        version: env!("CARGO_PKG_VERSION"),
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    /// The endpoint's whole job is to be non-empty and to carry both surfaces.
    /// An empty catalog is indistinguishable, on the client, from "this
    /// deployment has no such route" — which is why `oxyc` refuses one rather
    /// than serving it, and why the server must never produce one.
    #[tokio::test]
    async fn catalog_carries_both_surfaces_and_is_not_empty() {
        let Json(catalog) = get_catalog(Query(CatalogQuery::default())).await;

        assert!(
            catalog.routes.len() > 100,
            "the catalog shrank to {} routes — the build-time walk probably stopped following the router",
            catalog.routes.len()
        );
        assert!(
            catalog.routes.iter().any(|r| r.path.starts_with("/api/")),
            "no /api routes in the catalog"
        );
        assert!(
            catalog
                .routes
                .iter()
                .any(|r| r.path.starts_with("/external/api/")),
            "no /external/api routes in the catalog"
        );
        assert!(!catalog.surfaces.is_empty(), "no surfaces described");
    }

    /// Every route reports a role, because `oxyc routes` filters on it: the
    /// default view hides `ide-only` and `worker-only` mounts, which a caller
    /// hitting the load balancer cannot reach directly.
    #[tokio::test]
    async fn every_route_carries_a_role() {
        let Json(catalog) = get_catalog(Query(CatalogQuery::default())).await;
        for route in &catalog.routes {
            assert!(
                matches!(route.role, "fleet-ok" | "ide-only" | "worker-only"),
                "{} {} reported an unknown role {:?}",
                route.method,
                route.path,
                route.role
            );
        }
    }

    /// The catalog must describe ITSELF. A caller that cannot discover the
    /// discovery endpoint has to be told about it out of band, which is the
    /// situation this whole surface exists to remove.
    #[tokio::test]
    async fn the_catalog_lists_itself() {
        let Json(catalog) = get_catalog(Query(CatalogQuery::default())).await;
        assert!(
            catalog.routes.iter().any(|r| r.path == "/api/_catalog"),
            "/api/_catalog is missing from its own route table — the build-time \
             walk did not see its mount"
        );
    }

    /// `?filter=` narrows server-side, so a client after three routes does not
    /// download all ~670.
    #[tokio::test]
    async fn the_filter_narrows_the_table() {
        let Json(all) = get_catalog(Query(CatalogQuery::default())).await;
        let Json(filtered) = get_catalog(Query(CatalogQuery {
            filter: Some("threads".into()),
        }))
        .await;

        assert!(
            filtered.routes.len() < all.routes.len(),
            "the filter returned everything — it is not being applied"
        );
        assert!(!filtered.routes.is_empty(), "no route matched `threads`");
        // The haystack is built from EXACTLY the fields `search` reads, via the
        // same function it reads them with. Listing them here by hand made the
        // assertion a superset of the real filter, so it passed whether or not
        // `description` was searched — which is precisely how the endpoint's
        // doc came to claim a field the filter did not match.
        for route in &filtered.routes {
            let matched = route_catalog::routes()
                .iter()
                .find(|r| r.method == route.method && r.path == route.path)
                .map(route_catalog::searchable_fields)
                .expect("a filtered route is in the catalog")
                .iter()
                .any(|field| field.to_lowercase().contains("threads"));
            assert!(
                matched,
                "{} {} matched `threads` but contains it in none of the searchable fields",
                route.method, route.path
            );
        }
    }

    /// `description` really is searched, not merely documented as searched.
    ///
    /// The generic filter test cannot show this: `threads` appears in the path
    /// of every route it returns, so it would pass against a filter that
    /// ignored descriptions entirely. This needs a needle that lives ONLY in a
    /// description.
    #[tokio::test]
    async fn the_filter_reaches_descriptions() {
        // A word from some handler's doc comment that is in no path.
        let needle = route_catalog::routes()
            .iter()
            .filter(|r| !r.description.is_empty())
            .find_map(|r| {
                r.description
                    .split_whitespace()
                    .map(|w| {
                        w.trim_matches(|c: char| !c.is_alphanumeric())
                            .to_lowercase()
                    })
                    .find(|w| {
                        w.len() > 7
                            && route_catalog::routes()
                                .iter()
                                .all(|other| !other.path.to_lowercase().contains(w.as_str()))
                    })
            })
            .expect("some description carries a word that appears in no path");

        let Json(got) = get_catalog(Query(CatalogQuery {
            filter: Some(needle.clone()),
        }))
        .await;
        assert!(
            !got.routes.is_empty(),
            "`?filter={needle}` matched nothing — it only appears in a description, so \
             the filter is not reading descriptions"
        );
    }

    /// An empty or whitespace-only filter means "everything", not "nothing" —
    /// `?filter=` from a client that built the query string unconditionally
    /// must not come back as an empty catalog, which every caller reads as
    /// "this deployment has no routes".
    #[tokio::test]
    async fn a_blank_filter_is_not_a_filter() {
        let Json(all) = get_catalog(Query(CatalogQuery::default())).await;
        for blank in ["", "   "] {
            let Json(got) = get_catalog(Query(CatalogQuery {
                filter: Some(blank.into()),
            }))
            .await;
            assert_eq!(
                got.routes.len(),
                all.routes.len(),
                "a blank filter ({blank:?}) narrowed the table"
            );
        }
    }
}
